File: Algorithm.Python/IndustryStandardSecurityIdentifiersRegressionAlgorithm.py
# QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
# Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from AlgorithmImports import *

### <summary>
### Algorithm illustrating how to get a security's industry-standard identifier from its `Symbol`
### </summary>
class IndustryStandardSecurityIdentifiersRegressionAlgorithm(QCAlgorithm):

    def initialize(self):
        self.set_start_date(2014, 6, 5)
        self.set_end_date(2014, 6, 5)

        equity = self.add_equity("AAPL").symbol

        cusip = equity.cusip
        composite_figi = equity.composite_figi
        sedol = equity.sedol
        isin = equity.isin
        cik = equity.cik

        self.check_symbol_representation(cusip, "CUSIP")
        self.check_symbol_representation(composite_figi, "Composite FIGI")
        self.check_symbol_representation(sedol, "SEDOL")
        self.check_symbol_representation(isin, "ISIN")
        self.check_symbol_representation(f"{cik}", "CIK")

        # Check Symbol API vs QCAlgorithm API
        self.check_ap_is_symbol_representations(cusip, self.cusip(equity), "CUSIP")
        self.check_ap_is_symbol_representations(composite_figi, self.composite_figi(equity), "Composite FIGI")
        self.check_ap_is_symbol_representations(sedol, self.sedol(equity), "SEDOL")
        self.check_ap_is_symbol_representations(isin, self.isin(equity), "ISIN")
        self.check_ap_is_symbol_representations(f"{cik}", f"{self.cik(equity)}", "CIK")

        self.log(f"\n_aapl CUSIP: {cusip}"
                 f"\n_aapl Composite FIGI: {composite_figi}"
                 f"\n_aapl SEDOL: {sedol}"
                 f"\n_aapl ISIN: {isin}"
                 f"\n_aapl CIK: {cik}")

    def check_symbol_representation(self, symbol: str, standard: str) -> None:
        if not symbol:
            raise Exception(f"{standard} symbol representation is null or empty")

    def check_ap_is_symbol_representations(self, symbol_api_symbol: str, algorithm_api_symbol: str, standard: str) -> None:
        if symbol_api_symbol != algorithm_api_symbol:
            raise Exception(f"Symbol API {standard} symbol representation ({symbol_api_symbol}) does not match "
                            f"QCAlgorithm API {standard} symbol representation ({algorithm_api_symbol})")

File: Algorithm.CSharp/IndustryStandardSecurityIdentifiersRegressionAlgorithm.cs
/*
 * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
 * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
*/

using System;
using System.Collections.Generic;

using QuantConnect.Interfaces;
using QuantConnect.Util;

namespace QuantConnect.Algorithm.CSharp
{
    /// <summary>
    /// Algorithm illustrating how to get a security's industry-standard identifier from its <see cref="Symbol"/>
    /// </summary>
    public class IndustryStandardSecurityIdentifiersRegressionAlgorithm : QCAlgorithm, IRegressionAlgorithmDefinition
    {
        public override void Initialize()
        {
            SetStartDate(2014, 06, 05);
            SetEndDate(2014, 06, 05);

            var equity = AddEquity("AAPL").Symbol;

            var cusip = equity.CUSIP;
            var compositeFigi = equity.CompositeFIGI;
            var sedol = equity.SEDOL;
            var isin = equity.ISIN;
            var cik = equity.CIK;

            CheckSymbolRepresentation(cusip, "CUSIP");
            CheckSymbolRepresentation(compositeFigi, "Composite FIGI");
            CheckSymbolRepresentation(sedol, "SEDOL");
            CheckSymbolRepresentation(isin, "ISIN");
            CheckSymbolRepresentation($"{cik}", "CIK");

            // Check Symbol API vs QCAlgorithm API
            CheckAPIsSymbolRepresentations(cusip, CUSIP(equity), "CUSIP");
            CheckAPIsSymbolRepresentations(compositeFigi, CompositeFIGI(equity), "Composite FIGI");
            CheckAPIsSymbolRepresentations(sedol, SEDOL(equity), "SEDOL");
            CheckAPIsSymbolRepresentations(isin, ISIN(equity), "ISIN");
            CheckAPIsSymbolRepresentations($"{cik}", $"{CIK(equity)}", "CIK");

            Log($"\nAAPL CUSIP: {cusip}" +
                $"\nAAPL Composite FIGI: {compositeFigi}" +
                $"\nAAPL SEDOL: {sedol}" +
                $"\nAAPL ISIN: {isin}" +
                $"\nAAPL CIK: {cik}");
        }

        private static void CheckSymbolRepresentation(string symbol, string standard)
        {
            if (symbol.IsNullOrEmpty())
            {
                throw new Exception($"{standard} symbol representation is null or empty");
            }
        }

        private static void CheckAPIsSymbolRepresentations(string symbolApiSymbol, string algorithmApiSymbol, string standard)
        {
            if (symbolApiSymbol != algorithmApiSymbol)
            {
                throw new Exception($@"Symbol API {standard} symbol representation ({symbolApiSymbol}) does not match QCAlgorithm API {
                    standard} symbol representation ({algorithmApiSymbol})");
            }
        }

        /// <summary>
        /// This is used by the regression test system to indicate if the open source Lean repository has the required data to run this algorithm.
        /// </summary>
        public bool CanRunLocally { get; } = true;

        /// <summary>
        /// This is used by the regression test system to indicate which languages this algorithm is written in.
        /// </summary>
        public Language[] Languages { get; } = { Language.CSharp, Language.Python };

        /// <summary>
        /// Data Points count of all timeslices of algorithm
        /// </summary>
        public long DataPoints => 795;

        /// <summary>
        /// Data Points count of the algorithm history
        /// </summary>
        public int AlgorithmHistoryDataPoints => 0;

        /// <summary>
        /// This is used by the regression test system to indicate what the expected statistics are from running the algorithm
        /// </summary>
        public Dictionary<string, string> ExpectedStatistics => new Dictionary<string, string>
        {
            {"Total Orders", "0"},
            {"Average Win", "0%"},
            {"Average Loss", "0%"},
            {"Compounding Annual Return", "0%"},
            {"Drawdown", "0%"},
            {"Expectancy", "0"},
            {"Start Equity", "100000"},
            {"End Equity", "100000"},
            {"Net Profit", "0%"},
            {"Sharpe Ratio", "0"},
            {"Sortino Ratio", "0"},
            {"Probabilistic Sharpe Ratio", "0%"},
            {"Loss Rate", "0%"},
            {"Win Rate", "0%"},
            {"Profit-Loss Ratio", "0"},
            {"Alpha", "0"},
            {"Beta", "0"},
            {"Annual Standard Deviation", "0"},
            {"Annual Variance", "0"},
            {"Information Ratio", "0"},
            {"Tracking Error", "0"},
            {"Treynor Ratio", "0"},
            {"Total Fees", "$0.00"},
            {"Estimated Strategy Capacity", "$0"},
            {"Lowest Capacity Asset", ""},
            {"Portfolio Turnover", "0%"},
            {"OrderListHash", "d41d8cd98f00b204e9800998ecf8427e"}
        };
    }
}
